### PR TITLE
Safrole STF Test Vectors (GP Section 6) - Update

### PR DESCRIPTION
**Pull request by davxy**
### ⚠️ Temporarily closed in favor of https://github.com/w3f/jamtestvectors/pull/28 ⚠️

### CODE REVIEW

**Comment by davxy** [comment]
Added "Offenders" STF input  (ref GP 0.3.1, eq. 58, 59).
Currently not used by the provided vectors (some vectors will be provided soon)

**Comment by emielsebastiaan** [comment]
One suggestion for a complementary test vector is the case of including a set of 17 tickets during the lottery phase. 
This would cover GP-0.3.6-eq:74 (`|E_T|<=K`).
This would result in a not yet covered error.

**Comment by ggwpez** [comment]
Graymatter client passes all updated JSON tests (did not try the Codec stuff yet, but other CODEC stuff passes). Some comments:
- It looks like the tiny contest duration is 10 blocks, but i dont see this mentioned in the README.
- From what i understood, the keys that are passed to the ring setup need to use the padding point, but the ones that make it into the SAFROLE state should use zero.
- The punish tests revealed mistakes in our implementation. Thanks for adding it.

**Comment by ggwpez** [comment]
These tests also use `tickets_per_validator = 2` , or?  
It just occurred to me now that this implies that 2/3+1 majority does not lead to a successful contest anymore, since `(6*(2/3)+1) * 2 = 10`, missing two tickets.  
Maybe we should increase the tickets per validator to three for the tiny spec?  

I will post the final numbers onto the jamchain website to have the unambiguous for all future tests, currently tere is:

```elixir
config :tiny,
  validators: 6,
  slots_per_epoch: 12,
  ticket_submission_end_phase: 10

**Comment by dakk** [comment]
> Graymatter client passes all updated JSON tests (did not try the Codec stuff yet, but other CODEC stuff passes). Some comments:
> 
>     * From what i understood, the keys that are passed to the ring setup need to use the padding point, but the ones that make it into the SAFROLE state should use zero.

correct, in appending G it says "Note that in the case a key HB has no corresponding Bandersnatch point when constructing the ring, then the Bandersnatch padding point as stated by hosseini2024bandersnatch should be substituted."; so the padding point should be used only when constructing the ring

**Comment by greywolve** [comment]
Another suggestion for a complementary test is the case where there are useless tickets in `E_T`, i.e covering equation 6.35 (v0.5.0-3): `n ⊆ γ′a`

File: safrole/safrole.asn [hunk 1]
```diff
@@ -1,33 +1,57 @@
...
+    -- [ι] Validator keys and metadata to be drawn from next.
+    iota ValidatorsData,
+    -- [γ_a] Sealing-key contest ticket accumulator.
+    gamma-a TicketsAccumulator,
+    -- [γ_s] Sealing-key series of the current epoch.
+    gamma-s TicketsOrKeys,
+    -- [γ_z] Bandersnatch ring commitment.
+    gamma-z BandersnatchRingCommitment
+    -- [ψ_o'] Posterior offenders sequence.
+    post-offenders SEQUENCE OF Ed25519Public
```
Comment (ggwpez) [inline safrole/safrole.asn:32]: Why is there a posterior element in the state?

Comment (davxy) [reply safrole/safrole.asn:32]: I have received feedback in the chat suggesting that the input should include only elements found within the block, such as header items and extrinsics. The pre-state consists of elements already present in the state, which may or may not have been altered (e.g., prior, dagger, double dagger, posterior). All in all I found that it makes sense

Comment (davxy) [reply safrole/safrole.asn:32]: Nice catch, I'll fix it. Ty

Comment (davxy) [reply safrole/safrole.asn:32]: @ggwpez should be fixed

**Comment by davxy** [comment]
Temporarily closed in favor of https://github.com/w3f/jamtestvectors/pull/28

**Pull request closed without merging**